src/js/trove/table.js

```diff
@@ -13,10 +13,87 @@
     
     var brandTable = runtime.namedBrander("table", ["table: table brander"]);
     var annTable   = runtime.makeBranderAnn(brandTable, "Table");
 
+    var brandRow = runtime.namedBrander("row", ["table: row brander"]);
+    var annRow   = runtime.makeBranderAnn(brandRow, "Row");
+
+    var rowGetValue = runtime.makeMethod1(function(self, arg) {
+        ffi.checkArity(2, arguments, "get-value", false);
+        runtime.checkString(arg);
+        var index = self.$underlyingTable.headerIndex["column:" + arg];
+        if(typeof index === "number") {
+          return self.$rowData[index];
+        }
+        else {
+          return ffi.throwMessageException("No such column: " + arg);
+        }
+      });
+
+    var rowEquals = runtime.makeMethod2(function(self, other, rec) {
+        runtime.checkRow(self);
+        runtime.checkRow(other);
+        var headers1 = self.$underlyingTable.headerIndex;
+        var headers2 = other.$underlyingTable.headerIndex;
+        var hk1 = Object.keys(headers1);
+        var hk2 = Object.keys(headers2);
+        var rowData1 = self.$rowData;
+        var rowData2 = other.$rowData;
+        if(rowData1.length !== rowData2.length) {
+          return neq(self, other);
+        }
+        if(hk1.length !== hk2.length) {
+          return neq(self, other);
+        }
+        for(var i = 0; i < hk1.length; i += 1) {
+          if(headers1[hk1[i]] !== headers2[hk1[i]]) {
+            return neq(self, other);
+          }
+        }
+        return runtime.raw_array_fold(runtime.makeFunction(function(ans, val1, j) {
+          if (ffi.isNotEqual(ans)) { return ans; }
+          return runtime.safeCall(function() {
+            return rec.app(val1, rowData2[j]);
+          }, function(eqAns) {
+            return get(EQ, "equal-and").app(ans, eqAns);
+          }, "equals:combine-cells");
+        }), eq(), rowData1, 0);
+      });
+
+    function makeRow(underlyingTable, rowData) {
+      var rowVal = runtime.makeObject({
+        "get-value": rowGetValue,
+        "_equals": rowEquals
+      });
+      rowVal = applyBrand(brandRow, rowVal);
+      rowVal.$underlyingTable = underlyingTable;
+      rowVal.$rowData = rowData;
+      return rowVal;
+    }
+
+    function makeRowFromArray(rawArrayOfTuples) {
+      var headerIndex = [];
+      var rowData = [];
+
+      // TODO(joe): error checking here for bogus values, or elsewhere?
+      // May be good to keep this fast and do checks in Pyret-land
+      for(var i = 0; i < rawArrayOfTuples.length; i += 1) {
+        var colname = rawArrayOfTuples[i].vals[0];
+        if(headerIndex["column:" + colname] !== undefined) {
+          return runtime.ffi.throwMessageException("Duplicate column name in row: " + colname);
+        }
+        headerIndex["column:" + colname] = i;
+        rowData[i] = rawArrayOfTuples[i].vals[1];
+      }
+
+      return makeRow({ headerIndex: headerIndex }, rowData);
```

Comment:
Yeah – what I actually want to do (but didn't need to in order to make things work) is to construct a row out of a table and a raw array of values. The table will have an underlying array of headers, header index, and rows (which its methods close over now).

This would be a big performance win for programs that use lots of tables. For this PR, I just did the headerIndex object since that was all that was actually needed to implement row.